### PR TITLE
Add a .pragma directive for configuration files

### LINKED ISSUE #8207: VMS-specific bug in OpenSSL build configuration
Reported by Mark Berryman (<mark@theberrymans.com>)  on comp.os.vms; reproduced here.  This was derived from advice to OpenSSL builders on VMS platforms but should be considered as a refinement to the OpenSSL baseline.

Historically, OpenSSL has done symbol substitution in its configuration file as follows:
```
dir             = ssl1$root:[demoCA     # Where everything is kept
certs           = ${dir}.certs]         # Where the issued certs are kept
crl_dir         = ${dir}.crl]           # Where the issued crl are kept
database        = ${dir}]index.txt      # database index file.
new_certs_dir   = ${dir}.certs]         # default place for new certs.
certificate     = ${dir}]cacert.pem     # The CA certificate
serial          = ${dir}]serial.txt     # The current serial number
#crlnumber      = ${dir}]crlnumber      # the current crl number  - do we need .TXT or .?
crl             = ${dir}]crl.pem        # The current CRL
private_key     = ${dir}.private]cakey.pem      # The private key
```
Specify a dollar sign, a brace (or a parenthesis), the symbol name, and a closing brace.  At some point (somewhere in the 1.0.2 train it looks like) the OpenSSL team decided to make the enclosing braces optional. This means that if you have any dollar signs in your VMS file names (e.g. ssl1$root) then OpenSSL will try to treat the alphanumerics after the the $ as a symbol name and throw an error when it can't find it.

There are two ways to address this:
1. eliminate all dollar signs in your OpenSSL configuration file that are not used to indicate symbol substitution (such as replacing ssl1$root with a new logical name ssl1_root).
2. correct the code to make the enclosing braces once again mandatory.

The steps to do this are:
1. edit `[.crypto.conf]conf_def.c`
2. search for '$' (there should only be one instance)
3. change that line from
```
    else if (*from == '$') {
```
   to
```
#ifdef OPENSSL_SYS_VMS
     else if ((*from == '$') && ((from[1] == '{') || (from[1] == '('))) {
#endif
```
and rebuild OpenSSL.

A working configuration file is required for many OpenSSL operations. As you can see, I use the VMS-specific one distributed with SSL1.  I have used the above fix in OpenSSL 1.0.2o and later as well as in OpenSSL 1.1.1a.

### PR DESCRIPTION
**Pull request by levitte**
Currently added pragma:

.pragma dollarid:on

This allows dollar signs to be a keyword character unless it's
followed by a opening brace or parenthesis.

Fixes #8207

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
This should be fine to review at this point.

I'm pondering adding an escape character pragma as well, in another PR.  That could help Windows users that want to have the backslash reserved for paths...

**Comment by levitte** [comment]
Ping!

File: crypto/conf/conf_def.c [hunk 1]
```diff
@@ -354,7 +354,51 @@ static int def_load_bio(CONF *conf, BIO *in, long *line)
...
+                /*
+                 * Known pragmas:
+                 *
+                 * dollarid     takes "on", "true or "off", "false"
+                 */
+                if (strcmp(p, "dollarid") == 0) {
+                    if (strcmp(pval, "on") == 0
+                        || strcmp(pval, "true") == 0) {
+                        conf->flag_dollarid = 1;
+                        fprintf(stderr, "setting dollarid\n");
```
Comment (mattcaswell) [inline crypto/conf/conf_def.c]: This looks like debug code. Similarly below.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/conf/conf_def.c (hunk 1 above)
Comment (levitte) [reply crypto/conf/conf_def.c]: Yup.  Fixed

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

0255c1742a5a8186802fdef19b4f1c8c701a27bc Add a .pragma directive for configuration files
5aaba376189c0d606b832877b037d51caf338e7f Add a tester of OpenSSL config files and test the current .pragma

**Pull request closed without merging**